### PR TITLE
llvm: replace some deprecated functions

### PR DESCRIPTION
**Pull request by Luv-Ray**
`LLVMMDStringInContext` and `LLVMMDNodeInContext` are deprecated, replace them with `LLVMMDStringInContext2` and `LLVMMDNodeInContext2`.
Also replace `Value` with `Metadata` in some function signatures for better consistency.

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/llvm/ffi.rs [hunk 1]
```diff
@@ -912,12 +912,12 @@ unsafe extern "C" {
...
 
     // Operations on metadata
-    // FIXME: deprecated, replace with LLVMMDStringInContext2
+    #[deprecated = "replace with LLVMMDStringInContext2"]
     pub fn LLVMMDStringInContext(C: &Context, Str: *const c_char, SLen: c_uint) -> &Value;
 
     pub fn LLVMMDStringInContext2(C: &Context, Str: *const c_char, SLen: size_t) -> &Metadata;
 
-    // FIXME: deprecated, replace with LLVMMDNodeInContext2
+    #[deprecated = "replace with LLVMMDNodeInContext2"]
```
Comment (bjorn3) [inline compiler/rustc_codegen_llvm/src/llvm/ffi.rs:920]: Maybe just remove these extern declarations entirely if they aren't used anywhere?

Comment (Luv-Ray) [reply compiler/rustc_codegen_llvm/src/llvm/ffi.rs]: Done.

**Comment by nnethercote** [comment]
I unintentionally duplicated this effort in #130503. My PR looked extremley similar to this one, although I didn't add `Metadata` to `BackendTypes`. Anyway, looks good to me.

r? @nikic

File: compiler/rustc_codegen_llvm/src/builder.rs [hunk 2]
```diff
@@ -678,25 +679,21 @@ impl<'a, 'll, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
...
             ];
 
-            llvm::LLVMSetMetadata(
-                load,
-                llvm::MD_range as c_uint,
-                llvm::LLVMMDNodeInContext(self.cx.llcx, v.as_ptr(), v.len() as c_uint),
-            );
+            let md = llvm::LLVMMDNodeInContext2(self.cx.llcx, v.as_ptr(), v.len());
+            let md = llvm::LLVMMetadataAsValue(&self.llcx, md);
+            llvm::LLVMSetMetadata(load, llvm::MD_range as c_uint, md);
```
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/builder.rs:688]: Can we add a `set_metadata` method that does the LLVMMetadataAsValue + LLVMSetMetadata dance?

LLVM should really allow setting the metadata without first wrapping it in a value and then unwrapping again, but we can at least hide this...

File: compiler/rustc_codegen_llvm/src/context.rs [hunk 3]
```diff
@@ -586,6 +586,14 @@ impl<'ll, 'tcx> CodegenCx<'ll, 'tcx> {
             llvm::LLVMSetSection(g, c"llvm.metadata".as_ptr());
         }
     }
+
+    /// A wrapper for [`llvm::LLVMSetMetadata`], but it takes `Metadata` as a parameter instead of `Value`.
+    pub(crate) fn set_metadata<'a>(&self, val: &'a Value, kind_id: c_uint, md: &'a Metadata) {
```
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/context.rs:591]: Accept MetadataType here and move the c_uint cast into the method?

Comment (Luv-Ray) [reply compiler/rustc_codegen_llvm/src/context.rs]: Done :)

**Comment by nikic** [comment]
I can't even reproduce this when running the docker container...

@bors try

**Comment by nikic** [comment]
Okay, no idea what happened, so let's just try again...

@bors retry

**Comment by nikic** [comment]
Ah crap, I didn't realize that bors retry also retires try builds. I'll just let it run to avoid confusing bors further.

**Comment by nikic** [comment]
Huh, and now it failed again. So probably something is legitimately wrong with this patch that makes things break non-deterministically -- but I'm not able to spot it.

**Comment by nikic** [comment]
Okay, this test has been failing on some other PRs as well, I filed https://github.com/rust-lang/rust/issues/130656.

**Comment by nikic** [comment]
Problematic test has been removed, try again...

@bors r+

**Comment by Luv-Ray** [comment]
seems that it needs to run `retry`

**Pull request merged**